jgrapht-core/src/main/java/org/jgrapht/alg/util/Extension.java

```diff
@@ -34,60 +34,127 @@
  */
 package org.jgrapht.alg.util;
 
-import java.util.*;
+import java.util.HashMap;
+import java.util.Map;
 
 
 /**
- * Abstract extension manager allowing to extend given class-types with supplied
- * extension's class-type saving source class object references
+ * Abstract extension/encapsulation manager.
+ * This class creates and manages objects extensions and encapsulations. An object, from here on denoted as prototype,
+ * can be encapsulated in or extended by another object. An example would be the relation between an edge (prototype) and an annotated edge. The
+ * annotated edge encapsulates/extends an edge, thereby augmenting it with additional data.
+ * In symbolic form, if b is the prototype class, than a(b) would be its extension. This concept is similar to java's extension where
+ * one class is derived from (extends) another class (prototype).
  *
- * @param <T> class-type to be extended
- * @param <E> extension concept class-type
- */
+ * @param <T> class-type to be extended (class-type of prototype)
+ * @param <E> class-type of extension
+ *
+ *
+ *
+ * NOTE JK: This seems to be a poor name. This class is more of an ExtensionManager than an Extension by itself
+ * */
 public class Extension<T, E>
 {
+    /* Factory class to create new extensions */
     private ExtensionFactory<E> extensionFactory;
-    private Map<T, E> extensions = new HashMap<>();
+    /* Mapping of prototypes to their extensions */
+    private Map<T, E> prototypeToExtensionMap = new HashMap<>();
 
     public Extension(ExtensionFactory<E> factory)
     {
         this.extensionFactory = factory;
     }
 
+    /**
+     * Creates and returns an extension object.
+     * JK: Renamed this function to 'getInstance' for consistency with the getSingletonInstance function
+     * @return Extension object
+     */
+    @Deprecated
```

Comment:
First of all, let's stay consistent: previous naming scheme wasn't ideal, but current pick is even worse -- `getInstance` is a widespread convention for naming singleton-getters, therefore seeing `edgeExtensionManager.getInstance` hurts readability as hard as it could.

Much better choice would be to name those pair `createExtension`/`getExtension`.